File: Java/labs/lab22_Template_Method/task_2_1 – solution/src/Pizza.java
abstract public class Pizza {

    public abstract void cook();

    public void printInstructions() {  // шаблонний метод
        makeBase();
        addToppings();
        cook();
        System.out.println();
    }

    public void makeBase() {
        System.out.println("Mix flour, yeast and salt");
        System.out.println("Roll out the dough");
    }

    public void addToppings() {
        System.out.println("Add tomato");
        System.out.println("Add cheese");
    }

}

File: Cpp/lab22_Template_Method/task_2_1 - solution/src/pizza.cpp
#include <iostream>

#include "pizza.hpp"


void Pizza::makeBase()
{
  std::cout << "Mix flour, yeast and salt" << std::endl
            << "Roll out the dough" << std::endl;
}

void Pizza::addToppings()
{
  std::cout << "Add tomato" << std::endl
            << "Add cheese" << std::endl;
}

void Pizza::printInstructions()
{
  this->makeBase();
  this->addToppings();
  this->cook();
  std::cout << std::endl;
}
